File: app/src/main/java/com/example/doan/Utils/CartManager.java
package com.example.doan.Utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import com.example.doan.Models.CartItem;
import com.example.doan.Models.Product;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class CartManager {
    private static final String TAG = "CartManager";
    private static final String PREFS_NAME = "CartPrefs";
    private static final String KEY_CART_ITEMS = "cart_items";
    private static final String KEY_LAST_ACTIVITY = "last_activity_time";
    private static final long CART_TIMEOUT_MS = 5 * 60 * 1000; // 5 phút
    
    private static CartManager instance;
    private List<CartItem> cartItems;
    private Context context;
    private long lastActivityTime;
    private Gson gson;

    private CartManager() {
        cartItems = new ArrayList<>();
        gson = new Gson();
        lastActivityTime = System.currentTimeMillis();
    }

    public static synchronized CartManager getInstance() {
        if (instance == null) {
            instance = new CartManager();
        }
        return instance;
    }
    
    /**
     * Initialize CartManager with Context (call this in Application or first Activity)
     */
    public void init(Context context) {
        this.context = context.getApplicationContext();
        loadCart();
        loadLastActivityTime();
        clearCartIfExpired();
    }

    public void addToCart(Product product, int quantity, String sizeName, double unitPrice) {
        // Kiểm tra xem món này (cùng ID và cùng Size) đã có trong giỏ chưa
        for (CartItem item : cartItems) {
            if (item.getDrinkId() != null && item.getDrinkId() == product.getId() && 
                item.getSizeName() != null && item.getSizeName().equals(sizeName)) {
                item.setQuantity(item.getQuantity() + quantity);
                item.setTotalPrice(item.getUnitPrice() * item.getQuantity());
                updateLastActivityTime();
                saveCart();
                return;
            }
        }
        // Nếu chưa có, thêm mới
        CartItem newItem = new CartItem();
        newItem.setDrinkId((long) product.getId());
        newItem.setDrinkName(product.getName());
        newItem.setDrinkImage(product.getImageUrl());
        newItem.setSizeName(sizeName);
        newItem.setQuantity(quantity);
        newItem.setUnitPrice(unitPrice);
        newItem.setTotalPrice(unitPrice * quantity);
        
        cartItems.add(newItem);
        updateLastActivityTime();
        saveCart();
        Log.i(TAG, "Added to cart: " + product.getName() + " x" + quantity);
    }

    public void updateQuantity(int position, int newQuantity) {
        if (position >= 0 && position < cartItems.size()) {
            if (newQuantity <= 0) {
                cartItems.remove(position);
            } else {
                CartItem item = cartItems.get(position);
                item.setQuantity(newQuantity);
                item.setTotalPrice(item.getUnitPrice() * newQuantity);
            }
            updateLastActivityTime();
            saveCart();
        }
    }

    public void removeItem(int position) {
        if (position >= 0 && position < cartItems.size()) {
            cartItems.remove(position);
            updateLastActivityTime();
            saveCart();
            Log.i(TAG, "Removed item at position: " + position);
        }
    }

    public List<CartItem> getCartItems() {
        clearCartIfExpired();
        return cartItems;
    }

    public void clearCart() {
        cartItems.clear();
        saveCart();
        Log.i(TAG, "Cart cleared");
    }

    public double getTotalPrice() {
        double total = 0;
        for (CartItem item : cartItems) {
            total += item.getTotalPrice();
        }
        return total;
    }
    
    public int getItemCount() {
        return cartItems.size();
    }
    
    /**
     * Update last activity time (call when user interacts with cart)
     */
    public void updateLastActivityTime() {
        lastActivityTime = System.currentTimeMillis();
        saveLastActivityTime();
    }
    
    /**
     * Check if cart has expired (5 minutes of inactivity)
     */
    public void clearCartIfExpired() {
        long currentTime = System.currentTimeMillis();
        long timeDiff = currentTime - lastActivityTime;
        
        if (timeDiff > CART_TIMEOUT_MS) {
            Log.w(TAG, "Cart expired after " + (timeDiff / 1000) + " seconds. Clearing cart.");
            clearCart();
        }
    }
    
    /**
     * Save cart to SharedPreferences
     */
    private void saveCart() {
        if (context == null) {
            Log.e(TAG, "Context is null, cannot save cart");
            return;
        }
        
        try {
            SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
            String json = gson.toJson(cartItems);
            prefs.edit().putString(KEY_CART_ITEMS, json).apply();
            Log.d(TAG, "Cart saved: " + cartItems.size() + " items");
        } catch (Exception e) {
            Log.e(TAG, "Error saving cart: " + e.getMessage());
        }
    }
    
    /**
     * Load cart from SharedPreferences
     */
    private void loadCart() {
        if (context == null) {
            Log.e(TAG, "Context is null, cannot load cart");
            return;
        }
        
        try {
            SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
            String json = prefs.getString(KEY_CART_ITEMS, null);
            
            if (json != null && !json.isEmpty()) {
                Type type = new TypeToken<List<CartItem>>(){}.getType();
                cartItems = gson.fromJson(json, type);
                if (cartItems == null) {
                    cartItems = new ArrayList<>();
                }
                Log.d(TAG, "Cart loaded: " + cartItems.size() + " items");
            } else {
                cartItems = new ArrayList<>();
                Log.d(TAG, "No saved cart found");
            }
        } catch (Exception e) {
            Log.e(TAG, "Error loading cart: " + e.getMessage());
            cartItems = new ArrayList<>();
        }
    }
    
    /**
     * Save last activity time
     */
    private void saveLastActivityTime() {
        if (context == null) return;
        
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        prefs.edit().putLong(KEY_LAST_ACTIVITY, lastActivityTime).apply();
    }
    
    /**
     * Load last activity time
     */
    private void loadLastActivityTime() {
        if (context == null) {
            lastActivityTime = System.currentTimeMillis();
            return;
        }
        
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        lastActivityTime = prefs.getLong(KEY_LAST_ACTIVITY, System.currentTimeMillis());
    }
}

File: app/src/main/java/com/example/doan/Utils/CartManager.kt
package com.example.doan.Utils

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.example.doan.Models.CartItem
import com.example.doan.Models.Product
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

/**
 * ⚠️ WARNING: DEPRECATED - Có xung đột với Cart API từ server
 * 
 * Class này quản lý giỏ hàng LOCAL (SharedPreferences) nhưng app cũng sử dụng
 * Cart API từ server, gây ra mất đồng bộ dữ liệu nghiêm trọng.
 * 
 * KHUYẾN NGHỊ:
 * - Loại bỏ CartManager này hoàn toàn
 * - Chỉ sử dụng Cart API từ server (CartActivity đang dùng)
 * - Hoặc đồng bộ 2 chiều giữa local và server
 * 
 * HIỆN TẠI: Class này vẫn được giữ lại để tương thích với code cũ,
 * nhưng NÊN TRÁNH sử dụng trong code mới.
 */
class CartManager private constructor() {
    
    private var cartItems: MutableList<CartItem> = mutableListOf()
    private var context: Context? = null
    private var lastActivityTime: Long = System.currentTimeMillis()
    private val gson = Gson()
    
    fun init(context: Context) {
        this.context = context.applicationContext
        loadCart()
        loadLastActivityTime()
        clearCartIfExpired()
    }
    
    fun addToCart(product: Product, quantity: Int, sizeName: String, unitPrice: Double) {
        // Kiểm tra xem món này (cùng ID và cùng Size) đã có trong giỏ chưa
        cartItems.find { 
            it.drinkId == product.id.toLong() && it.sizeName == sizeName 
        }?.let { item ->
            item.quantity = (item.quantity ?: 0) + quantity
            item.totalPrice = (item.unitPrice ?: 0.0) * (item.quantity ?: 0)
            updateLastActivityTime()
            saveCart()
            return
        }
        
        // Nếu chưa có, thêm mới
        val newItem = CartItem(
            drinkId = product.id.toLong(),
            drinkName = product.name,
            drinkImage = product.imageUrl,
            sizeName = sizeName,
            quantity = quantity,
            unitPrice = unitPrice,
            totalPrice = unitPrice * quantity
        )
        
        cartItems.add(newItem)
        updateLastActivityTime()
        saveCart()
        Log.i(TAG, "Added to cart: ${product.name} x$quantity")
    }
    
    fun updateQuantity(position: Int, newQuantity: Int) {
        if (position in cartItems.indices) {
            if (newQuantity <= 0) {
                cartItems.removeAt(position)
            } else {
                cartItems[position].apply {
                    quantity = newQuantity
                    totalPrice = (unitPrice ?: 0.0) * newQuantity
                }
            }
            updateLastActivityTime()
            saveCart()
        }
    }
    
    fun removeItem(position: Int) {
        if (position in cartItems.indices) {
            cartItems.removeAt(position)
            updateLastActivityTime()
            saveCart()
            Log.i(TAG, "Removed item at position: $position")
        }
    }
    
    fun getCartItems(): List<CartItem> {
        clearCartIfExpired()
        return cartItems
    }
    
    fun clearCart() {
        cartItems.clear()
        saveCart()
        Log.i(TAG, "Cart cleared")
    }
    
    fun getTotalPrice(): Double {
        return cartItems.sumOf { it.totalPrice ?: 0.0 }
    }
    
    fun getItemCount(): Int = cartItems.size
    
    fun updateLastActivityTime() {
        lastActivityTime = System.currentTimeMillis()
        saveLastActivityTime()
    }
    
    fun clearCartIfExpired() {
        val currentTime = System.currentTimeMillis()
        val timeDiff = currentTime - lastActivityTime
        
        if (timeDiff > CART_TIMEOUT_MS) {
            Log.w(TAG, "Cart expired after ${timeDiff / 1000} seconds. Clearing cart.")
            clearCart()
        }
    }
    
    private fun saveCart() {
        context?.let { ctx ->
            try {
                val prefs = ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                val json = gson.toJson(cartItems)
                prefs.edit().putString(KEY_CART_ITEMS, json).apply()
                Log.d(TAG, "Cart saved: ${cartItems.size} items")
            } catch (e: Exception) {
                Log.e(TAG, "Error saving cart: ${e.message}")
            }
        } ?: Log.e(TAG, "Context is null, cannot save cart")
    }
    
    private fun loadCart() {
        context?.let { ctx ->
            try {
                val prefs = ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                val json = prefs.getString(KEY_CART_ITEMS, null)
                
                if (!json.isNullOrEmpty()) {
                    val type = object : TypeToken<MutableList<CartItem>>() {}.type
                    cartItems = gson.fromJson(json, type) ?: mutableListOf()
                    Log.d(TAG, "Cart loaded: ${cartItems.size} items")
                } else {
                    cartItems = mutableListOf()
                    Log.d(TAG, "No saved cart found")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error loading cart: ${e.message}")
                cartItems = mutableListOf()
            }
        } ?: Log.e(TAG, "Context is null, cannot load cart")
    }
    
    private fun saveLastActivityTime() {
        context?.let { ctx ->
            val prefs = ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            prefs.edit().putLong(KEY_LAST_ACTIVITY, lastActivityTime).apply()
        }
    }
    
    private fun loadLastActivityTime() {
        context?.let { ctx ->
            val prefs = ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            lastActivityTime = prefs.getLong(KEY_LAST_ACTIVITY, System.currentTimeMillis())
        } ?: run {
            lastActivityTime = System.currentTimeMillis()
        }
    }
    
    companion object {
        private const val TAG = "CartManager"
        private const val PREFS_NAME = "CartPrefs"
        private const val KEY_CART_ITEMS = "cart_items"
        private const val KEY_LAST_ACTIVITY = "last_activity_time"
        private const val CART_TIMEOUT_MS = 5 * 60 * 1000L // 5 phút
        
        @Volatile
        private var instance: CartManager? = null
        
        fun getInstance(): CartManager {
            return instance ?: synchronized(this) {
                instance ?: CartManager().also { instance = it }
            }
        }
    }
}
